sshd-common/src/main/java/org/apache/sshd/common/cipher/ChaChaCipher.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.sshd.common.cipher;
+
+import java.util.Arrays;
+
+import javax.crypto.AEADBadTagException;
+import javax.crypto.Cipher;
+import javax.crypto.SecretKey;
+import javax.crypto.spec.IvParameterSpec;
+import javax.crypto.spec.SecretKeySpec;
+
+import org.apache.sshd.common.mac.BaseMac;
+import org.apache.sshd.common.mac.Mac;
+import org.apache.sshd.common.util.buffer.BufferUtils;
+import org.apache.sshd.common.util.security.SecurityUtils;
+
+/**
+ * AEAD cipher based on the
+ * <a href="https://github.com/openbsd/src/blob/master/usr.bin/ssh/PROTOCOL.chacha20poly1305">OpenSSH ChaCha20-Poly1305</a>
+ * cipher extension.
+ */
+public class ChaChaCipher extends BaseCipher {
+    protected Mode mode;
+    protected SecretKey bodyKey;
+    protected SecretKey headerKey;
+    protected Cipher bodyCipher;
+    protected Cipher headerCipher;
+    protected boolean initialized;
+    protected PacketSequenceNumberParameterSpec parameters;
+
+    protected final Mac poly1305 = new BaseMac("Poly1305", 16, 32, true);
+    protected final byte[] macKey = new byte[32];
+
+    public ChaChaCipher() {
+        super(8, 16, 64, "ChaCha", 256, "ChaCha", 8);
+    }
+
+    @Override
+    protected Cipher createCipherInstance(Mode mode, byte[] key, byte[] iv) throws Exception {
+        this.mode = mode;
```

Comment:
Recommend adding `this.mode = Objects.requireNotNull(mode, "No cipher mode provided
);`

Reply:
To match the superclass behavior, this should actually be changed to default to `Mode.Decrypt` when `mode` is `null`.